libs/vkd3d/va_map.c

```diff
@@ -0,0 +1,381 @@
+
+    if (block)
+    {
+        if (va < vkd3d_atomic_uint64_load_explicit(&block->l.va, vkd3d_memory_order_relaxed))
+            resource = vkd3d_atomic_ptr_load_explicit(&block->l.resource, vkd3d_memory_order_relaxed);
+        else if (va >= vkd3d_atomic_uint64_load_explicit(&block->r.va, vkd3d_memory_order_relaxed))
+            resource = vkd3d_atomic_ptr_load_explicit(&block->r.resource, vkd3d_memory_order_relaxed);
+    }
+
+    if (!resource)
+    {
+        pthread_mutex_lock(&va_map->mutex);
+        resource = vkd3d_va_map_find_small_entry(va_map, va, NULL);
+        pthread_mutex_unlock(&va_map->mutex);
+    }
+
+    return resource;
+}
+
+VkDeviceAddress vkd3d_va_map_alloc_fake_va(struct vkd3d_va_map *va_map, VkDeviceSize size)
+{
+    struct vkd3d_va_allocator *allocator = &va_map->va_allocator;
+    struct vkd3d_va_range range;
+    VkDeviceAddress va;
+    size_t i;
+    int rc;
+
+    if ((rc = pthread_mutex_lock(&allocator->mutex)))
+    {
+        ERR("Failed to lock mutex, rc %d.\n", rc);
+        return 0;
+    }
+
+    memset(&range, 0, sizeof(range));
+
+    /* The free list is ordered in such a way that the largest range
+     * is always first, so we don't have to iterate over the list */
+    if (allocator->free_range_count)
+        range = allocator->free_ranges[0];
+
+    if (range.size >= size)
+    {
+        va = range.base;
+
+        range.base += size;
+        range.size -= size;
+
+        for (i = 0; i < allocator->free_range_count - 1; i++)
+        {
+            if (allocator->free_ranges[i + 1].size > range.size)
+                allocator->free_ranges[i] = allocator->free_ranges[i + 1];
+            else
+                break;
+        }
+
+        if (range.size)
+            allocator->free_ranges[i] = range;
+        else
+            allocator->free_range_count--;
+    }
+    else
+    {
+        va = allocator->next_va;
+        allocator->next_va += size;
+    }
+
+    pthread_mutex_unlock(&allocator->mutex);
+    return va;
+}
+
+void vkd3d_va_map_free_fake_va(struct vkd3d_va_map *va_map, VkDeviceAddress va, VkDeviceSize size)
+{
+    struct vkd3d_va_allocator *allocator = &va_map->va_allocator;
+    size_t range_idx, range_shift, i;
+    struct vkd3d_va_range new_range;
+    int rc;
+
+    if ((rc = pthread_mutex_lock(&allocator->mutex)))
+    {
+        ERR("Failed to lock mutex, rc %d.\n", rc);
+        return;
+    }
+
+    new_range.base = va;
+    new_range.size = size;
+
+    range_idx = allocator->free_range_count;
+    range_shift = 0;
+
+    /* Find and effectively delete any free range adjacent to new_range */
+    for (i = 0; i < allocator->free_range_count; i++)
+    {
+        const struct vkd3d_va_range *cur_range = &allocator->free_ranges[i];
+
+        if (range_shift)
+            allocator->free_ranges[i - range_shift] = *cur_range;
+
+        if (cur_range->base == va + size || cur_range->base + cur_range->size == va)
```

Comment:
Shouldn't this check against new_range.base/size instead of va/size? I can imagine a case where we have 3 adjacent ranges, A, B, C. If A and C are in the free list, and we free B, we should get two separate merge events I think?

Reply:
Shouldn't really change anything since we can't have more than one adjacent range with a lower base address, for example, unless the merging has already gone wrong elsewhere. Will change anyway since aligning the fake VAs does get messy otherwise.